pulp_smash/tests/puppet/cli/test_sync_from_puppet_skip_duplicate_content.py

```diff
@@ -0,0 +1,70 @@
+# coding=utf-8
+"""Tests that to verify Puppet sync on duplicate contents."""
+from __future__ import unicode_literals
+
+import unittest2
+
+from pulp_smash import cli, config, utils
+from pulp_smash.constants import PUPPET_FEED, PUPPET_QUERY
+from pulp_smash.tests.puppet.utils import set_up_module as setUpModule  # noqa pylint:disable=unused-import
+
+
+class SyncFromPuppetSkipDuplicateContentTestCase(unittest2.TestCase):
+    """Test whether the importer recognizes the content already in Pulp.
+
+    According to `Pulp #1937`_, Pulp sometimes fails to sync units into a
+    repository if the unit already existed.
+
+    How do we test this? The test steps would be as follows:
+
+    1. upload or sync a module to repo1
+    2. perform a sync on repo2 against a forge-style repository
+        (one with a modules.json file) that includes the module from 1.
+    3. Assert the module from 1. is in the published repo2.
+        (i.e., just check if the 2nd sync would succeed)
+    Please see reference from: https://github.com/PulpQE/pulp-smash/issues/269
+
+    To do this in pulp-admin, it could be something as simple as:
+
+    1. pulp-admin puppet repo create --repo-id=repo1
+        --feed=http://forge.puppetlabs.com --queries=torssh
+    2. pulp-admin puppet repo sync run --repo-id=repo1
+    3. pulp-admin puppet repo create --repo-id=repo2
+        --feed=http://forge.puppetlabs.com --queries=torssh
+    4. pulp-admin puppet repo sync run --repo-id=repo2
+
+    The end result is a test case that syncs an existing Puppet modules
+    into two repositories and check whether the 2nd succeeds.
+
+    .. _Pulp #1937: https://pulp.plan.io/issues/1937
+    """
+
+    def test_sync_puppet_forge(self):
+        """Create two Puppet repositories and trigger their syncs."""
+        cfg = config.get_config()
+        utils.pulp_admin_login(cfg)
+
+        # Create two different repositories and schedule them for syncs.
+        repo_id1 = utils.uuid4()
+        repo_id2 = utils.uuid4()
+        if repo_id2 == repo_id1:
+            repo_id2 = utils.uuid4()
+
+        client = cli.Client(cfg)
+        for repo_id in (repo_id1, repo_id2):
+            # Create the repository.
+            cmd = (
+                'pulp-admin puppet repo create '
+                '--repo-id {} --feed {} --queries {}'
+            ).format(repo_id, PUPPET_FEED, PUPPET_QUERY)
+            client.run(cmd.split())
+            # Sync the repository.
+            cmd = (
+                'pulp-admin puppet repo sync run --repo-id {}'
+            ).format(repo_id)
+            client.run(cmd.split())
+
+        # Delete the two repos.
+        for repo_id in (repo_id1, repo_id2):
+            cmd = 'pulp-admin puppet repo delete --repo-id {}'.format(repo_id)
+            self.addCleanup(client.run, cmd.split())
```

Comment:
What happens if an error occurs while executing `pulp-admin puppet repo sync`? In that case, these `addCleanup` functions will never execute. `addCleanup` should run immediately after each resource is created.